### ISSUE TITLE
SSRS 2017 custom authentication not working

### ISSUE THREAD

**Issue by genghis2011** [Microsoft/Reporting-Services#84]
Hello,
I am not able to start ssrs 2017 service with custom authentication. Any help are appreciated.

I have a windows 10 which already has sql server 2016 and ssrs 2016 installed, and ssrs 2016 is configured to use the custom authentication.  I installed ssrs 2017 only on the same windows 10(no sql sever 2017 database engine). 
I followed this link to set up ssrs 2017 custom authentication: https://github.com/Microsoft/Reporting-Services/tree/master/CustomSecuritySample. 
I added <MachineKey> to RSReportServer.config file. When I tried to start ssrs 2017 from Report server configuration manager, it didn’t start. 
RSHostingService_xxx.log has the below error:
Exception occurred while configuring and starting processesSystem.UriFormatException: Invalid URI: The hostname could not be parsed.
   at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind)
   at System.UriBuilder..ctor(String uri)
   at Microsoft.BIServer.Configuration.ExternalUrlHelpers.CalculateUrl(Application application, String externalReportServerUrl, String externalUrlRoot, Int32 secureConnectionLevel)
   at Microsoft.BIServer.BIService.BIService.UpdateManagedProcessConfigs(ServiceConfig serviceConfig, ManagementState managementState)
   at Microsoft.BIServer.BIService.BIService.<ReloadHostingState>d__12.MoveNext()

Per https://github.com/Microsoft/Reporting-Services/issues/26, the issue seems to be machine key issue. I followed the steps mentioned to generate the machine key, and verified that the casing of the attributes is Pascal Casing. But I still got the same error message in RSHostingService log.

**Comment by manceelee** [comment]
I think it has to do something with UrlString setting on RSreportserverconfig.  Try to delete all <URL> under <URLs>, and restart the server.

**Comment by genghis2011** [comment]
Hi manceelee, There are two <<UrlString> fields, one is for ReportServer, another is for Reports. 
<UrlString>http://+:87</UrlString>
"Try to delete all under" on your suggestion, I am not sure what you mean.

**Comment by manceelee** [comment]
Make copy of your rsreportserver
Remove URL portion under URLS in rsreportserver.config, remove all Binding under Bindings 
Remove the machine key for now, you can add it back later.  
restart the server! 

For adding a new machine key, please follow the same steps in 
https://github.com/Microsoft/Reporting-Services/issues/49#issuecomment-367299570

**Comment by genghis2011** [comment]
Hi manceelee, I removed URL portion under URLS and MachineKey fields in rsreportserver.config. Since it was not configured to use SSL, there are no bindings field.  
I tried to restart SSRS server, it didn't work. RSHostingService log has the same error. 
The resreportserver.config file has following on <URLReservations>: 
	<URLReservations>
		<Application>
			<Name>ReportServerWebService</Name>
			<VirtualDirectory>ReportServer</VirtualDirectory>
			<URLs>
	                </URLs>
		</Application>
		<Application>
			<Name>ReportServerWebApp</Name>
			<VirtualDirectory>Reports</VirtualDirectory>
			<URLs>
			</URLs>
		</Application>
	</URLReservations>

**Comment by manceelee** [comment]
Hi genghis2011, please restart the the window server, not only SSRS server.  80% of time restart the server fix my issue, please give it a try.

**Comment by manceelee** [comment]
also please remove anything inside UrlRoot and ReportServerUrl

**Comment by genghis2011** [comment]
manceelee, I removed anything inside UrlRoot and ReportServerUrl and restart the windows. SSRS server is able to start and I can also access ReportServer url and login with form authentication.
 However, Web portal url doesn't work, it gave HTTP 500 error. RSPortal.log has below error message
SSRS 2017 uses the same custom authentication dll which worked with SSRS 2016. By looking at CustomSecuritySample and CustomSecuritySample2016, I don't see there are any changes on GetUserInfo() between 2017 and 2016. 
Do you have any idea on the error? Thank you for the help.

Exception: Microsoft.BIServer.Owin.Common.Exceptions.AuthenticationExtensionException: GetUserInfo ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.BIServer.Owin.Common.Services.AuthenticationService.<>c__DisplayClass1.<GetUserInfo>b__0()
   at Microsoft.BIServer.Owin.Common.Services.AuthenticationService.WrapExtensionCall[T](String methodName, Func`1 func)
   --- End of inner exception stack trace ---
   at Microsoft.BIServer.Owin.Common.Services.AuthenticationService.WrapExtensionCall[T](String methodName, Func`1 func)
   at Microsoft.BIServer.Owin.Common.Middleware.CustomAuthenticationMiddleware.CreatePortalIdentity(IOwinContext context)
   at Microsoft.BIServer.Owin.Common.Middleware.CustomAuthenticationMiddleware.Invoke(IOwinContext context)
   at Microsoft.Owin.Security.Infrastructure.AuthenticationMiddleware`1.<Invoke>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.BIServer.Owin.Common.Middleware.RequestLoggingMiddleWare.<Invoke>d__2.MoveNext() | RequestID = s_d22123cc-d2ad-4893-b09b-c249c5706b50 
RSPortal!reportserverwebapp!RSPortal.exe!3!04/19/2018-14:34:04:: i INFO: Configuration changed. Restarting ReportServerWebApp
RSPortal!reportserverwebapp!RSPortal.exe!3!04/19/2018-14:34:04:: i INFO: Starting ReportServerWebApp
RSPortal!reportserverwebapp!RSPortal.exe!3!04/19/2018-14:34:04:: i INFO: Working directory : C:\Program Files\Microsoft SQL Server Reporting Services\SSRS\Portal
RSPortal!reportserverwebapp!RSPortal.exe!3!04/19/2018-14:34:04:: i INFO: Report Server url: http://d9p6mlc2:87/ReportServer
RSPortal!reportserverwebapp!RSPortal.exe!3!04/19/2018-14:34:04:: i INFO: Report Server Web App virtual root: /Reports
RSPortal!reportserverwebapp!RSPortal.exe!3!04/19/2018-14:34:04:: i INFO: Web app url: http://+:87/Reports/
RSPortal!reportserverwebapp!RSPortal.exe!3!04/19/2018-14:34:04:: i INFO: Authentication scheme(s): Anonymous

**Comment by manceelee** [comment]
have you added back the machine key, post your rsreportserverconfig and let me take a look.

**Comment by genghis2011** [comment]
Yes, I added machine key back. I did not add anything under ReportServerUrl. 
When using IIS to generate the key, I tried with selecting check-box of "Automatically generate at runtime", and also tried with un-checking it. 
Here is the config file:
[rsreportserver.txt](https://github.com/Microsoft/Reporting-Services/files/1929498/rsreportserver.txt)

**Comment by manceelee** [comment]
try again
replace the following machinekey to your config

<machineKey validationKey="3AD264B8B9C7AEE932A81EF4F34910AAAE5A800D2E8A6B776EF5D770F2FD1E7EE511B8875E63D0E72D4F268D33596049309ACDEE303A7C1683AF2FDFB927DEBE" decryptionKey="C45F99DE02F06BB7D37CD3757DCEF73622F98DB2DCBB5027" validation="AES" decryption="AES" />

Open Report Server Configuration Manager
Go to Encryption keys
Delete Encrypted content

**Comment by genghis2011** [comment]
The machinekey didn't work. 
I attached the debug to process. It failed on Setconfiguration()--- > doc.LoadXml() in AuthenticationExtension.cs
        public void SetConfiguration(String configuration)
        {
            var doc = new XmlDocument(); 
####            doc.LoadXml(configuration);      

The value of "configuration" string is "root2E0CF8BFA97508A37E67E480665D856BE917" on ssrs2017. doc.LoadXml() couldn't load the xml string. 
This value is supposed to be the configuration element under Authentication element in rsreportserverconfig file. e.g:
![configuration](https://user-images.githubusercontent.com/38299375/39021420-b703940c-43fe-11e8-9505-9d6eeaee3d67.PNG)

**Comment by gboreki** [comment]
@genghis2011 If you are not getting the configuration as an Xml string this sounds like a bug in SSRS, given that the sample here does not have a configuration in the authentication extension. Can you place you configuration in a separate file and read from that to see if it workaround the problem for you?

Thanks
-G

**Comment by genghis2011** [comment]
@gboreki, it looks like a bug on ssrs 2017. The same codes work on ssrs 2016. I will use a workaround to get the configuration data. 

 I have another problem on using visual studio to build dll. 
When I tried to build dll, it gave below error message:
CS7027	Error signing output with public key from file '..\..\samplekey.snk' -- The process cannot access the file because it is being used by another process.
I couldn't find another process using the key file. I even stopped the ssrs and restart windows server. 
I didn't have this issue when I built the dll for ssrs 2016 last year. 
Thanks in advance for any help.
